packages/alfa-dom/src/node/document.ts

```diff
@@ -125,6 +127,30 @@ export namespace Document {
       Node.fromNode(child, device),
     ).map((children) => Document.of(children, json.style.map(Sheet.from)));
   }
+
+  export function cloneDocument(
+    document: Document,
+    newElements: Element[],
+    predicate: Predicate<Element>,
+    device?: Device,
+  ): Trampoline<Document> {
+    return Trampoline.traverse(document.children(), (child) => {
+      if (Element.isElement(child) && predicate(child)) {
+        return Trampoline.done(newElements);
+      }
+
+      return Node.cloneNode(child, newElements, predicate, device).map(
+        (node) => [node],
+      );
```

Comment:
Regarding the suggestion, I think the challenge is that the mapper should return some kind of `Trampoline<U>` so I think we do need to lift the single `node` into `[node]` inside the `.map()` function, otherwise we end up with an array of trampolines it seems 🤔

Regarding the iterators, might be possible, but not totally sure how well it would play together with `Trampoline`